Adding MQTT module(client)

DanielBallaSZTE opened the pull request:
This patch adds the MQTT client support to IoT.js.

Co-authored-by: Daniel Vince <vinced@inf.u-szeged.hu>
IoT.js-DCO-1.0-Signed-off-by: Daniel Balla dballa@inf.u-szeged.hu

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,584 @@
+  offset += sizeof(variable_header_protocol);
+
+  // Flags
+  memcpy(buffer_wrap->buffer + offset, &connect_flag, sizeof(connect_flag));
```
No need `memcpy` here, simply `buffer_wrap->buffer[offset] = connect_flags`

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,584 @@
+  jerry_value_t jkeepalive =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_KEEPALIVE);
+
+  uint8_t connect_flag = 0;
```
plural: flags

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,584 @@
+  switch (packet_type) {
+    case CONNACK:
+      if (remaining_length != 2) {
+        return JS_CREATE_ERROR(COMMON, "MQTT: CONNACK packet is corrupted");
```
Maybe we should think about closing the connection.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,584 @@
+
+  char first_byte = buffer[0];
+  char remaining_length = buffer[1];
+  char packet_type = (first_byte >> 4) & 0b00001111;
```
0x0f

zherczeg commented on src/modules/iotjs_module_mqtt.h:
```diff
@@ -0,0 +1,241 @@
+
+enum {
+  // Reserved bit, must be 0
+  FLAG_RESERVED = 1 << 0,
```
MQTT_FLAG_....

zherczeg commented on src/modules/iotjs_module_mqtt.h:
```diff
@@ -0,0 +1,241 @@
+    uint8_t msb;
+    uint8_t lsb;
+    char *buffer;
+  } data;
```
Do we need this struct in struct?

DanielBallaSZTE replied:
No, we don't thanks for pointing out.

DanielBallaSZTE commented:
@zherczeg I've changed those things, thank you for your time!

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,677 @@
+
+static void iotjs_mqtt_calculate_msb_lsb(uint16_t len, uint8_t *msb,
+                                         uint8_t *lsb) {
+  *msb = (uint8_t)((len & 0xFF00) >> 8);
```
no need `& 0xFF00`

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,677 @@
+  uint8_t *buff_ptr = (uint8_t *)buffer_wrap->buffer;
+
+  // Fixed header
+  *buff_ptr++ = header_byte;
```
I think this is the typical case when you don't need comment. It is obvious what you do here.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,677 @@
+ * Waits for a uint32_t remaining_length.
+ * Returns the encoded value with the correct size.
+ */
+static char *iotjs_encode_remaining_length(uint32_t len) {
```
I feel this is overcomplicated. You could write the data directly into the stream.

DanielBallaSZTE replied:
We could, but we need the size of that data to allocate the buffer.

DanielBallaSZTE replied:
I figured out a solution for this.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,677 @@
+  // It is an uint32_t at maximum
+  char temp_buff[4];
+  do {
+    char d = len % 128;
```
`len & 0x7F;`

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,677 @@
+  char temp_buff[4];
+  do {
+    char d = len % 128;
+    len /= 128;
```
`len >>= 7`

zherczeg commented on src/modules/iotjs_module_mqtt.h:
```diff
@@ -0,0 +1,243 @@
+ * - Specific flags to each MQTT Control Packet.
+ */
+typedef struct {
+#ifdef MQTT_REVERSED
```
How can this be reversed???? There is no bit order, only byte order.

DanielBallaSZTE replied:
This is left over from some old code, we never use it, it's going to get deleted when we clean up the header file.

DanielBallaSZTE commented:
@zherczeg I have updated the PR, thank you!

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,777 @@
+
+  do {
+    if (++length > IOTJS_MODULE_MQTT_MAX_REMAINING_LENGTH_BYTES) {
+      return JS_CREATE_ERROR(COMMON, "MQTT: The arrived message is too big");
```
You should return with a special value to show error. E.g. `UINT32_MAX`, since the message size is limited to 256 Mbyte.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,777 @@
+}
+
+
+static void iotjs_mqtt_calculate_msb_lsb(uint16_t len, uint8_t *msb,
```
In general I would do the following for such functions:
```
static uint8_t* iotjs_mqtt_encode_length(uint8_t *buffer, uint16_t len)
{
  buffer[0] = (uint8_t)(len >> 8);
  buffer[1] = (uint8_t)(len & 0xff);
  return buffer + 2;
}
```
So they copy their arguments into the stream and advances the string pointer.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,777 @@
+  *buff_ptr++ = header_byte;
+  *buff_ptr++ = remaining_length;
+  *buff_ptr++ = packet_identifier_msb;
+  *buff_ptr++ = packet_identifier_lsb;
```
For the last bytes you can use `buff_ptr[0]` - `buff_ptr[3]`. The last ++ has no effect anyway.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,777 @@
+  iotjs_mqtt_payload_t topic_payload =
+      iotjs_create_payload(iotjs_bufferwrap_from_jbuffer(jtopic));
+
+  uint8_t header_byte = 0;
```
`uint8_t header_byte = (UNSUBSCRIBE << 4);`
No need every step in separeted lines.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,777 @@
+      sizeof(packet_identifier_msb) + sizeof(packet_identifier_lsb);
+  uint32_t remaining_length = payload_len + variable_header_len;
+  size_t full_len = sizeof(header_byte) + payload_len + variable_header_len +
+                    get_remaining_length_size(remaining_length);
```
I would compute these in message order:
`sizeof(header_byte) + get_remaining_length_size(remaining_length) + variable_header_len + payload_len`

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,777 @@
+
+  *buff_ptr++ = header_byte;
+
+  buff_ptr += iotjs_encode_remaining_length(remaining_length, buff_ptr);
```
if `iotjs_encode_remaining_length` returns with buffer, no need `+=`. The `=` is always faster since the compiler does not need to preserve the value of `buffer`.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,777 @@
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_QOS);
+
+  uint8_t qos = 0;
+  if (!jerry_value_is_undefined(jqos) && !jerry_value_has_error_flag(jqos)) {
```
I think there is a utility now which converts strings and buffers to char*, which is exactly what you need here. It also returns with error if the conversion is not possible.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,777 @@
+  char *buffer = buffer_wrap->buffer;
+
+  char first_byte = buffer[0];
+  char packet_type = (first_byte >> 4) & 0b00001111;
```
`0x0f`. Please don't use 0b

DanielBallaSZTE commented:
@zherczeg Thank you again for the review, I have modified lots of things, and removed some unused things from the header file. What we still need are some checks if the user gives the correct input.
Besides from that, I'm open for any suggestions :)

zherczeg commented on src/js/mqtt.js:
```diff
@@ -0,0 +1,308 @@
+    keepalive: { value: options.keepalive || 60 },
+  });
+
+  this._socket = new net.Socket();
```
Is it a simple `net.Socket` even if tls is specified? I would remove the _protocol member and would simply accept a duplex stream as options.socket. If it is not specified, a new `net.Socket` is created.

zherczeg commented on src/js/mqtt.js:
```diff
@@ -0,0 +1,308 @@
+  this._socket.on('end', this.ondisconnect);
+};
+
+MQTTClient.prototype.disconnect = function(error) {
```
I think `end` is more common than `disconnect`.

DanielBallaSZTE replied:
There actually is a packet type which has the name `DISCONNECT`, causing the server to properly disconnect the client.

zherczeg commented on src/js/mqtt.js:
```diff
@@ -0,0 +1,292 @@
+  this._socket.write(buf);
+}
+
+MQTTClient.prototype.onpingresp = function() {
```
Let me rephrase the question. These seem internal functions. However they are exposed through `MQTTClient.prototype`. What about simply create a `function onpingresp() {...}`. This function would be not visible externally, but could use anything internally.

zherczeg commented on src/js/mqtt.js:
```diff
@@ -0,0 +1,308 @@
+  if (!Buffer.isBuffer(connectionOptions.clientId)) {
+    connectionOptions.clientId = new Buffer(connectionOptions.clientId);
+  }
+  return MQTTClient(connectionOptions);
```
`return new MQTTClient(connectionOptions);`

DanielBallaSZTE replied:
Yeah about the prototype thing, we haven't refactored the JS code yet, were focusing on the C side. 
We were planning on changing them as you mentioned it.

zherczeg commented on src/modules/iotjs_module_mqtt.h:
```diff
@@ -0,0 +1,130 @@
+enum {
+  QoS0 = 0b00, // At most once delivery.
+  QoS1 = 0b01, // At least once delivery.
+  QoS2 = 0b10  // Exactly once delivery.
```
Please use 0x, or here you can use decimal numbers.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,719 @@
+
+
+static uint8_t *iotjs_encode_remaining_length(uint32_t len,
+                                              unsigned char *buffer) {
```
`uint8_t *buffer`. Usually buffer (= destination byte-stream pointer) is always the first argument for such functions.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,719 @@
+}
+
+
+static uint32_t iotjs_decode_remaining_length(char *buffer, size_t *offset) {
```
Since `get_remaining_length_size` is called earlier, I would move the error checking there, and just use some asserts here. Perhaps get_remaining_length_size could return 0 on error, since at least 1` byte is required if the encoding is possible.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,719 @@
+    }
+    c = (unsigned char)buffer[*offset];
+    remaining_length += (c & 127) * multiplier;
+    multiplier *= 128;
```
`multiplier <<= 7`

zherczeg replied:
Btw usually shift is faster than multiplication, so you could simply use a shift:
```
remaining_length += (c & 0x7F) << shift;
shift += 7
```
Also `|=` can be used instead of `+=` here, but they are equally fast so they don't provide any advantage.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,719 @@
+
+static void iotjs_mqtt_calculate_length(uint16_t *len, uint8_t msb,
+                                        uint8_t lsb) {
+  *len = (msb << 8) | lsb;
```
What about `static uint16_t iotjs_mqtt_calculate_length(uint8_t msb, uint8_t lsb) { ... }`?

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,719 @@
+}
+
+
+static uint8_t *iotjs_mqtt_string_serialize(uint8_t *dst_buffer,
```
For me encode/serialize is equally ok, but please only use on of them. E.g `iotjs_encode_remaining_length` -> `iotjs_mqtt_serialize_remaining_length` or `iotjs_mqtt_string_encode`. Perhaps encode sounds a bit better, but either of them is ok.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,719 @@
+  }
+
+
+  iotjs_bufferwrap_t *client_id = iotjs_bufferwrap_from_jbuffer(jclient_id);
```
No type check here. I mentioned that there is a function which creates iotjs string from strings/buffers, or return NULL on error.

DanielBallaSZTE replied:
I don't think it is needed since JS side a buffer is always created for the client id, and that's passed to the C side.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,719 @@
+  uint8_t packet_identifier_msb = 0;
+
+  // IF NOT QOS
+  if (!jerry_value_is_undefined(jqos) &&
```
There isa  to_boolean function, which returns with `bool` (never throws error). I would use it here.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,719 @@
+      break;
+    }
+    case DISCONNECT:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Broker disconnected the client");
```
Why this is an error? Isn't it a notification?

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,719 @@
+    case CONNECT:
+    case SUBSCRIBE:
+    case UNSUBSCRIBE:
+    case PINGREQ:
```
Server cannot send ping?

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,719 @@
+      header.bits.type = PUBLISH;
+      header.bits.dup = first_byte & 0x08;
+      header.bits.qos = (first_byte & 0x06) >> 1;
+      header.bits.retain = first_byte & 0x01;
```
I think more checks would be useful here. E.g. jerry offers utf8 validating functions.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,719 @@
+    }
+    case PUBACK: {
+      if (remaining_length != 2) {
+        return JS_CREATE_ERROR(COMMON, "MQTT: PUBACK packet is corrupted");
```
Is the connection is aborted on error? I think the standard expects it.

zherczeg commented on src/modules/iotjs_module_mqtt.c:
```diff
@@ -0,0 +1,719 @@
+}
+
+
+JS_FUNCTION(MqttHandle) {
```
What happens if the package is not fully arrived?

vincedani replied:
The maximum TCP package size is 64K, but due to memory limitations the IoT.js can't handle such a big packages.

zherczeg commented:
Travis: `Signed-off-by message is incorrect. The following line should be at the end of the e5135d6 commit's message: 'IoT.js-DCO-1.0-Signed-off-by: Daniel Vince vinced@inf.u-szeged.hu'. `

DanielBallaSZTE commented:
@daeyeon I've updated the commit! Thank you

yichoi commented:
LGTM
